### ISSUE TITLE
`Invariant` (with certain invariants) makes certain patterns (ie `IndexMut`) impossible to implement

### ISSUE THREAD

**Issue by sarsko** [xldenis/creusot#794]
Here exemplified with a `Heap` which implements `IndexMut<usize>`. 
The source of the conflict is
```rust
&& forall<i: Int> 0 <= i && i < self.heap@.len() ==> 
       self.heap@[i]@ < self.indices@.len()
```
in the `invariant` (as we have no way to guarantee that whoever gets the mutable borrow does not violate this property upon mutating the `&mut usize` returned by `IndexMut`).

```rust
extern crate creusot_contracts;
use creusot_contracts::invariant::{Invariant, self};
use creusot_contracts::{*, pearlite, ensures, requires, predicate, trusted, vec};
use ::std::ops::{Index, IndexMut};

struct Heap {
    heap: Vec<usize>,
    indices: Vec<usize>,
}

impl Invariant for Heap {
    #[predicate]
    fn invariant(self) -> bool {
        pearlite!{
            self.heap@.len() <= self.indices@.len()
            && forall<i: Int> 0 <= i && i < self.heap@.len() ==> 
               self.heap@[i]@ < self.indices@.len()
        }
    }
}

impl Index<usize> for Heap {
    type Output = usize;
    #[inline]
    #[requires(ix@ < self.heap@.len())]
    #[ensures(self.heap@[ix@] == *result)]
    fn index(&self, ix: usize) -> &usize {
        &self.heap[ix]
    }
}

impl IndexMut<usize> for Heap {
    #[inline]
    #[requires(ix@ < self.heap@.len())]
    #[ensures((*self).heap@[ix@] == *result)]
    #[ensures((^self).heap@[ix@] == ^result)]
    #[ensures(forall<i: Int> 0 <= i && i != ix@ && i < self.heap@.len() ==> self.heap@[i] == (^self).heap@[i])]
    #[ensures((^self).heap@.len() == (*self).heap@.len())]
    fn index_mut(&mut self, ix: usize) -> &mut usize {
        &mut self.heap[ix]
    }
}
```

**Comment by xldenis** [comment]
> in the invariant (as we have no way to guarantee that whoever gets the mutable borrow does not violate this property upon mutating the &mut usize returned by IndexMut).

Indeed, this is the motivating use case for `after_expiry` style mechanisms (like Prusti). @voidc, this might be a usecase you're interested in?

However, even without `Invariant`, how could you soundly implement `IndexMut`? What is preventing someone from breaking your heap property?

**Comment by sarsko** [comment]
> However, even without `Invariant`, how could you soundly implement `IndexMut`? What is preventing someone from breaking your heap property?

Well, without `Invariant`, soundly implementing `IndexMut` is trivial. Under `Invariant` the invariant belongs to the data structure, whereas it belongs to the caller when we manually enforce it. The caller may thus break the heap property at will, and will then lose it until it is reestablished.

Some loose ideas on the matter:
1. I would like for things like `Invariant` to be the way to do things. It makes proofs substantially shorter (and thus more readable) and removes one of the most common causes of proofs not checking out, thus making the proof writer more effective. If ie giving out a mutable borrow to a part of a structure which has an `Invariant` is impossible or super cumbersome, then that would be very unfortunate.
2. I initially thought that what should happen is that there should be a "contract" on the mutable borrow which upon expiry, requires that heap invariant should be established. This is better than what exists now, but is still a bit unflexible. For instance, iterating over the `heap` and doing updates to it, and then afterwards updating the `indices` is not possible under such a scheme.
3. One could say that the public API of a data structure should maintain its invariants, and it is thus not sound to offer `IndexMut` (or any other function which gives a mutable borrow to a part of its structure) for any structure where the types are not restrictive enough to enforce the invariant by themselves. However, currently, if we `pub` the fields of the structure, we allow callers to break (and establish) the invariants on their own. It just happens that this is disallowed if it passes through a function boundary. It would be nice to have a way to provide the same functionality while using functions / without `pub`ing fields. Allowing functions to transfer responsibility of the invariant to the caller would also be useful for private functions which manipulate `self` while it is not satisfying its invariant (ie during initialization).
4. I think simply having a way to state that a function opts out of the invariant should be enough to solve this issue. The function does not require the invariant, and it does not ensure the invariant, meaning that responsibility of the invariant is moved to the caller. This corresponds to the way things are done when manually enforcing invariants (but opt-out rather than opt-in).
